controllers/mirroring/mirroring_controller.go

```diff
@@ -0,0 +1,477 @@
+			return ctrl.Result{}, nil
+		}
+		r.log.Error(err, "Failed to get StorageClusterPeer.")
+		return ctrl.Result{}, err
+	}
+
+	if storageClusterPeer.Status.State != ocsv1.StorageClusterPeerStatePeered {
+		return ctrl.Result{}, fmt.Errorf(
+			"waiting for StorageClusterPeer %s to be in Peered state",
+			storageClusterPeer.Name,
+		)
+	}
+
+	ocsClient, err := providerClient.NewProviderClient(r.ctx, storageClusterPeer.Spec.ApiEndpoint, util.OcsClientTimeout)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create a new provider client: %v", err)
+	}
+	defer ocsClient.Close()
+
+	if err = r.reconcileRbdMirror(clientMappingConfig); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if res, err := r.reconcileBlockPoolMirroring(
+		ocsClient,
+		clientMappingConfig,
+		storageClusterPeer,
+		true,
+	); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	if res, err := r.reconcileRadosNamespaceMirroring(
+		ocsClient,
+		clientMappingConfig,
+		storageClusterPeer); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MirroringReconciler) reconcileRbdMirror(clientMappingConfig *corev1.ConfigMap) error {
+	rbdMirror := &rookCephv1.CephRBDMirror{}
+	rbdMirror.Name = util.CephRBDMirrorName
+	rbdMirror.Namespace = clientMappingConfig.Namespace
+
+	storageConsumers := &ocsv1alpha1.StorageConsumerList{}
+	if err := r.list(
+		storageConsumers,
+		client.MatchingFields{util.AnnotationIndexName: util.RequestMaintenanceModeAnnotation},
+	); err != nil {
+		return fmt.Errorf("failed to list StorageConsumer(s): %v", err)
+	}
+
+	if len(storageConsumers.Items) > 1 {
+		if err := r.delete(rbdMirror); err != nil {
+			return fmt.Errorf("failed to delete CephRBDMirror: %v", err)
+		}
+	}
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, rbdMirror, func() error {
+		if err := r.own(clientMappingConfig, rbdMirror); err != nil {
+			return err
+		}
+		rbdMirror.Spec.Count = 1
+		return nil
+	})
+	if err != nil {
+		r.log.Error(err, "Failed to create/update the CephRBDMirror", "CephRBDMirror", rbdMirror)
+		return err
+	}
+
+	return nil
+}
+
+func (r *MirroringReconciler) reconcileBlockPoolMirroring(
+	ocsClient *providerClient.OCSProviderClient,
+	clientMappingConfig *corev1.ConfigMap,
+	storageClusterPeer *ocsv1.StorageClusterPeer,
+	enableMirroring bool,
+) (ctrl.Result, error) {
+
+	selector := labels.NewSelector()
+	blockPoolLabelSelectorRequirement, err := labels.NewRequirement(
+		util.ForbidMirroringLabel,
+		selection.NotEquals,
+		[]string{"true"},
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	selector = selector.Add(*blockPoolLabelSelectorRequirement)
+
+	cephBlockPoolsList := &rookCephv1.CephBlockPoolList{}
+	if err = r.list(
+		cephBlockPoolsList,
+		client.InNamespace(clientMappingConfig.Namespace),
+		client.MatchingLabelsSelector{Selector: selector},
+	); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to list cephBlockPools: %v", err)
+	}
```

Comment:
I prefer if you list everything and then skip the labeled one in your for loop. This will create a more maintainable and readable code